pravega/pravega
# Issue 6221: Setting BookieID for every bookie instance

Signed-off-by: SrishT <Srishti.Thakkar@dell.com>

**Change log description**  
Sets a unique `bookieId` value for every newly created bookkeeper instance, and reads the persisted `bookieId` value for every other bookkeeper instance, so that the client can easily distinguish between an instance which is being newly created vs an instance which might have been restarted for some reason.

**Purpose of the change**  
Fixes #6221 

**What the code does**  
Sets a unique `bookieId` value for every newly created bookkeeper instance and mounts its value for every newly added bookkeeper instance to a persistent volume, so that it can be easily accessed even if this instance happens to get restarted at some point. 
However, in case this is not a fresh bookkeeper cluster installation, and we are upgrading from a previous bookkeeper version (which does not provide users the support to explicitly configure the BookieID) to the bookkeeper version which provides this support, the code reuses the older nomenclature for the BookieID, so that we do not face an `InvalidCookieException` due to a mismatch in the BookieID of any bookeeper instance post upgrade. In addition to this, a teardown script is added which would be invoked (only if  we are on the latest bookkeeper operator), and would delete the cookies for the bookie that is being scaled down at the time of pod deletion.

**How to verify it**  
Following are the various use cases that have been verified with I/O running in the background
- old version of bookkeeper operator
  old version of bookkeeper
  upgraded bookkeeper to latest version (computes BookieID from hostname and persists it in a file within a journal subdirectory)
  upgraded bookkeeper operator to latest version
  (this approach uses older nomenclature for BookieID as expected, and uses new nomenclature for any new bookkeeper pod that comes up as a result of any scale up)

- old version of bookkeeper operator
  old version of bookkeeper
  upgraded bookkeeper operator to latest version
  upgraded bookkeeper to latest version
  (this approach uses older nomenclature for BookieID as expected, and uses new nomenclature for any new bookkeeper pod that comes up as a result of any scale up)

- latest version of bookkeeper operator (i.e. one which contains [this](https://github.com/pravega/bookkeeper-operator/pull/166) fix)
  old version of bookkeeper
  upgraded bookkeeper to latest version
  (this approach uses older nomenclature for BookieID as expected, and uses new nomenclature for any new bookkeeper pod that comes up as a result of any scale up)

- latest version of bookkeeper operator (i.e. one which contains [this](https://github.com/pravega/bookkeeper-operator/pull/166) fix)
  latest version of bookkeeper
  deletion of pod should cause it to read the BookieID from the persisted file
  scale down followed by a scale up will not find any BookieID stored and will therefore generate a new BookieID and persist it
  (this approach uses new nomenclature for BookieID as expected) 

Also verified that we are getting a green system test build over the last multiple runs.
